v3/lints/rfc/lint_key_usage_and_extended_key_usage_inconsistent.go

```diff
@@ -0,0 +1,195 @@
+				return &lint.LintResult{Status: lint.Error}
+			}
+		}
+	}
+	return &lint.LintResult{Status: lint.Pass}
+}
+
+func init() {
+	lint.RegisterLint(&lint.Lint{
+		Name:          "e_key_usage_and_extended_key_usage_inconsistent",
+		Description:   "The certificate MUST only be used for a purpose consistent with both key usage extension and extended key usage extension.",
+		Citation:      "RFC 5280, Section 4.2.1.12.",
+		Source:        lint.RFC5280,
+		EffectiveDate: util.RFC5280Date,
+		Lint:          &KUAndEKUInconsistent{},
+	})
+}
+
+//  CheckConsistencyWithEKU* functions return false if the certificate being linted
+//  has inconsistent Key Usage bits set with a specific Extended Key Usage
+
+//CheckConsistencyWithEKUServerAuth checks if KU bits are consistent with Server Authentication EKU bit
+func CheckConsistencyWithEKUServerAuth(c *x509.Certificate) bool {
+	//  RFC 5280 4.2.1.12 on KU consistency with Server Authentication EKU:
+	//    -- TLS WWW server authentication
+	//    -- Key usage bits that may be consistent: digitalSignature,
+	//    -- keyEncipherment or keyAgreement
+
+	allowedCombinations := []x509.KeyUsage{
+		x509.KeyUsageDigitalSignature,
+		x509.KeyUsageKeyEncipherment,
+		x509.KeyUsageKeyAgreement,
+	}
+
+	return containsCombination(c.KeyUsage, allowedCombinations)
+}
+
+//CheckConsistencyWithEKUClientAuth checks if KU bits are consistent with Client Authentication EKU bit
+func CheckConsistencyWithEKUClientAuth(c *x509.Certificate) bool {
+	// 	RFC 5280 4.2.1.12 on KU consistency with Client Authentication EKU:
+	//    -- TLS WWW client authentication
+	//    -- Key usage bits that may be consistent: digitalSignature
+	//    -- and/or keyAgreement
+
+	allowedCombinations := []x509.KeyUsage{
+		x509.KeyUsageDigitalSignature,
+		x509.KeyUsageKeyAgreement,
+		x509.KeyUsageDigitalSignature | x509.KeyUsageKeyAgreement}
+
+	return containsCombination(c.KeyUsage, allowedCombinations)
+}
+
+//CheckConsistencyWithEKUCodeSigning checks if KU bits are consistent with Code Signing EKU bit
+func CheckConsistencyWithEKUCodeSigning(c *x509.Certificate) bool {
+	// 	RFC 5280 4.2.1.12 on KU consistency with Code Signing EKU:
+	//   -- Signing of downloadable executable code
+	//   -- Key usage bits that may be consistent: digitalSignature
+
+	allowedCombinations := []x509.KeyUsage{
+		x509.KeyUsageDigitalSignature}
+
+	return containsCombination(c.KeyUsage, allowedCombinations)
+}
+
+//CheckConsistencyWithEKUEmailProtection checks if KU bits are consistent with Email Protection EKU bit
+func CheckConsistencyWithEKUEmailProtection(c *x509.Certificate) bool {
+	// 	RFC 5280 4.2.1.12 on KU consistency with Email Protection EKU:
+	// 	  -- Email protection
+	//    -- Key usage bits that may be consistent: digitalSignature,
+	//    -- nonRepudiation, and/or (keyEncipherment or keyAgreement)
+	//  Note: Recent editions of X.509 have renamed nonRepudiation bit to contentCommitment
+
+	allowedCombinations := []x509.KeyUsage{
```

Comment:
Of course, Go doesn't have sets natively (which is a bit absurd imho) but the idiomatic alternative tends to be a `map[<T>]bool` which I think would work well in these cases as it obviates the need for helper function.

```suggestion
	return map[x509.KeyUsage]bool{
		x509.KeyUsageDigitalSignature:                                                                 true,
		x509.KeyUsageContentCommitment:                                                                true,
		x509.KeyUsageKeyEncipherment:                                                                  true,
		x509.KeyUsageKeyAgreement:                                                                     true,
		x509.KeyUsageDigitalSignature | x509.KeyUsageContentCommitment:                                true,
		x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment:                                  true,
		x509.KeyUsageDigitalSignature | x509.KeyUsageKeyAgreement:                                     true,
		x509.KeyUsageContentCommitment | x509.KeyUsageKeyEncipherment:                                 true,
		x509.KeyUsageContentCommitment | x509.KeyUsageKeyAgreement:                                    true,
		x509.KeyUsageDigitalSignature | x509.KeyUsageContentCommitment | x509.KeyUsageKeyEncipherment: true,
		x509.KeyUsageDigitalSignature | x509.KeyUsageContentCommitment | x509.KeyUsageKeyAgreement:    true,
	}[c.KeyUsage]
```